File: 0234-palindrome-linked-list/0234-palindrome-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    ListNode mid(ListNode head){
        ListNode f=head;
        ListNode s=head;
        while(f!=null&&f.next!=null){
            f=f.next.next;
            s=s.next;
        }
        return s;
    }
    ListNode rev(ListNode head){
        if(head==null||head.next==null){
            return head;
        }
        ListNode prev=null;
        ListNode curr=head;
        while(curr!=null){
            ListNode n=curr.next;
            curr.next=prev;
            prev=curr;
            curr=n;
        }
        return prev;
    }
    public boolean isPalindrome(ListNode head) {
        ListNode middle=mid(head);
        ListNode head2=rev(middle);
        ListNode temp=head;
        while(temp!=null&&head2!=null){
            if(temp.val!=head2.val){
                break;
            }
            temp=temp.next;
            head2=head2.next;
        }
        return head2==null||temp==null;
    }
}

File: 0234-palindrome-linked-list/0234-palindrome-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rev(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev = None
        curr = head
        
        while curr:
            nxt = curr.next     
            curr.next = prev    
            prev = curr         
            curr = nxt          
        
        return prev
    def middle(self,head):
        f=head
        s=head
        while f and f.next:
            f=f.next.next
            s=s.next
        return s

    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        if not head or not head.next:
            return True
        middle=self.middle(head)
        h2=self.rev(middle)
        t=head
        while  t and h2:
            if t.val!=h2.val:
                return False
            t=t.next
            h2=h2.next
        return True

        